service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,297 @@
+                    oldGiftCertificate.setName(newGiftCertificateName);
+                } else {
+                    errorMessage.append("Gift certificate with this name has already.");
+                }
+            }
+            if (checkGiftCertificationDescription(newGiftCertificateDescription, errorMessage)) {
+                oldGiftCertificate.setDescription(newGiftCertificateDescription);
+            }
+            if (checkGiftCertificationPrice(newPrice, errorMessage)) {
+                oldGiftCertificate.setPrice(newPrice);
+            }
+            if (checkGiftCertificationDuration(newDuration, errorMessage)) {
+                oldGiftCertificate.setDuration(newDuration);
+            }
+            if (newTags != null) {
+                for (Tag newTag : newTags) {
+                    long newTagId = newTag.getId();
+                    checkTagName(newTag.getName(), errorMessage);
+                    if (newTagId != 0) {
+                        Optional<Tag> daoTagOptional = tagDao.findById(newTagId);
+                        if (!daoTagOptional.isPresent()) {
+                            errorMessage.append("Invalid tag id.Not founded.");
+                        }
+                    }
+                }
+            }
+            if (errorMessage.length() == 0) {
+                oldGiftCertificate.setLastUpdateDate(LocalDateTime.now());
+                giftCertificateDao.update(oldGiftCertificate);
+                if (newTags != null) {
+                    for (Tag newTag : newTags) {
+                        if (newTag.getId() != 0) {
+                            tagDao.update(newTag);
+                        } else {
+                            oldGiftCertificate.setId(id);
+                            connectGiftCertificateAndTag(oldGiftCertificate, newTag);
+                        }
+                    }
+                }
+            } else {
+                throw new InvalidFieldValueException(errorMessage.toString());
+            }
+        } else {
+            throw new ResourceNotFoundedException("Requested resource not found (id)=" + id);
+        }
+    }
+
+    private void connectGiftCertificateAndTag(GiftCertificate giftCertificate, Tag tag) {
+        long id = giftCertificate.getId();
+        long daoTagId;
+        Optional<Tag> daoTagOptional = tagDao.findByName(tag.getName());
+        if (daoTagOptional.isPresent()) {
+            Tag daoTag = daoTagOptional.get();
+            List<Tag> oldTagList = giftCertificate.getTags();
+            boolean matchResult = oldTagList.stream().noneMatch(t -> t.getName().equals(daoTag.getName()));
+            if (matchResult) {
+                daoTagId = daoTag.getId();
+                giftCertificatesTagDao.add(id, daoTagId);
+            }
+        } else {
+            daoTagId = tagDao.add(tag);
+            giftCertificatesTagDao.add(id, daoTagId);
+        }
+    }
+
+    @Override
+    public GiftCertificate findById(long id) {
```

Comment:
The service layer should return DTO objects instead of entity objects. So that we are able not to reveal the way we store objects in the DB (internal entity) to user, as we return DTO object instead that includes only the fields that are intended to be returned for the user. Please rework all service methods where needed.